### PR TITLE
Fix LlamaDynamicNTKScalingRotaryEmbedding cache

### LINKED ISSUE #25306: "Dynamic" Issue in LlamaDynamicNTKScalingRotaryEmbedding - Long context inference will impact short context inference.
### System Info

- `transformers` version: 4.32.0.dev0
- Platform: Linux-5.15.109+-x86_64-with-glibc2.35
- Python version: 3.10.12
- Huggingface_hub version: 0.16.4
- Safetensors version: 0.3.1
- Accelerate version: 0.22.0.dev0
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.0.1+cu118 (True)
- Tensorflow version (GPU?): 2.12.0 (True)
- Flax version (CPU?/GPU?/TPU?): 0.7.0 (gpu)
- Jax version: 0.4.13
- JaxLib version: 0.4.13
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help?

@sgugger 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

Please see my colab code:  
https://colab.research.google.com/drive/1SnQQxW7WMHgSOvAwF_HIlIDrAuXZ4IKp?usp=sharing

I asked the same prompt twice, with a long-context prompt inserted in between. However, this intermediate long-context inference resulted in different answers for the same question before and after it.

### Expected behavior

Since the input length of the tested prompts is within the maximum input token capacity the model can handle, the significance of "Dynamic" lies in ensuring that the embeddings for the inputs before and after remain the same, and consequently, the output results should also be the same.

I reviewed the code of the class "[LlamaDynamicNTKScalingRotaryEmbedding](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L147C8-L147C8)" and I think that due to caching, when the model infers a long context, the cached values of `cos_cached` and `sin_cached` are updated to adapt to the longer context. This causes the issue when the model infers a shorter context again.

### PR DESCRIPTION
**Pull request by i4never**
# What does this PR do?

Fixes #25306 . Cached pos embed will only be used when cached length is `model.config.max_position_embeddings` and asked `seq_len` is small or equal than this.
It's a little bit slower but correct. Value of sin/cos could be calculated in first decoder layer and passed to others (that's what [TGI flash_llama_modeling](https://github.com/huggingface/text-generation-inference/blob/f9910d13e296989f41e714c43eb60ce051359db3/server/text_generation_server/models/custom_modeling/flash_llama_modeling.py#L443-L460) does)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. #25306 
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?
@gante 

## Test
```python
import time
import random
import torch
from transformers.models.llama.modeling_llama import LlamaRotaryEmbedding


class LlamaDynamicNTKScalingRotaryEmbedding(LlamaRotaryEmbedding):
    """LlamaRotaryEmbedding extended with Dynamic NTK scaling. Credits to the Reddit users /u/bloc97 and /u/emozilla"""

    def __init__(self, dim, max_position_embeddings=2048, base=10000, device=None, scaling_factor=1.0):
        self.scaling_factor = scaling_factor
        super().__init__(dim, max_position_embeddings, base, device)

    def _set_cos_sin_cache(self, seq_len, device, dtype):
        if seq_len <= self.max_position_embeddings:
            seq_len = self.max_position_embeddings

        self.max_seq_len_cached = seq_len

        if seq_len > self.max_position_embeddings:
            base = self.base * (
                    (self.scaling_factor * seq_len / self.max_position_embeddings) - (self.scaling_factor - 1)
            ) ** (self.dim / (self.dim - 2))
        else:
            base = self.base
        inv_freq = 1.0 / (base ** (torch.arange(0, self.dim, 2).float().to(device) / self.dim))

        t = torch.arange(self.max_seq_len_cached, device=device, dtype=inv_freq.dtype)

        freqs = torch.einsum("i,j->ij", t, inv_freq)
        # Different from paper, but it uses a different permutation in order to obtain the same calculation
        emb = torch.cat((freqs, freqs), dim=-1)
        self.register_buffer("cos_cached", emb.cos()[None, None, :, :].to(dtype), persistent=False)
        self.register_buffer("sin_cached", emb.sin()[None, None, :, :].to(dtype), persistent=False)

    def forward(self, x, seq_len=None):
        # x: [bs, num_attention_heads, seq_len, head_size]
        if not (seq_len <= self.max_seq_len_cached == self.max_position_embeddings):
            self._set_cos_sin_cache(seq_len=seq_len, device=x.device, dtype=x.dtype)

        return (
            self.cos_cached[:, :, :seq_len, ...].to(dtype=x.dtype),
            self.sin_cached[:, :, :seq_len, ...].to(dtype=x.dtype),
        )


class LlamaDynamicNTKScalingRotaryEmbeddingOld(LlamaRotaryEmbedding):
    """LlamaRotaryEmbedding extended with Dynamic NTK scaling. Credits to the Reddit users /u/bloc97 and /u/emozilla"""

    def __init__(self, dim, max_position_embeddings=2048, base=10000, device=None, scaling_factor=1.0):
        self.scaling_factor = scaling_factor
        super().__init__(dim, max_position_embeddings, base, device)

    def _set_cos_sin_cache(self, seq_len, device, dtype):
        self.max_seq_len_cached = seq_len

        if seq_len > self.max_position_embeddings:
            base = self.base * (
                    (self.scaling_factor * seq_len / self.max_position_embeddings) - (self.scaling_factor - 1)
            ) ** (self.dim / (self.dim - 2))
            inv_freq = 1.0 / (base ** (torch.arange(0, self.dim, 2).float().to(device) / self.dim))
            self.register_buffer("inv_freq", inv_freq, persistent=False)

        t = torch.arange(self.max_seq_len_cached, device=device, dtype=self.inv_freq.dtype)

        freqs = torch.einsum("i,j->ij", t, self.inv_freq)
        # Different from paper, but it uses a different permutation in order to obtain the same calculation
        emb = torch.cat((freqs, freqs), dim=-1)
        self.register_buffer("cos_cached", emb.cos()[None, None, :, :].to(dtype), persistent=False)
        self.register_buffer("sin_cached", emb.sin()[None, None, :, :].to(dtype), persistent=False)


rope_ntk_new = LlamaDynamicNTKScalingRotaryEmbedding(128)
rope_ntk_old = LlamaDynamicNTKScalingRotaryEmbeddingOld(128)

x = torch.rand(1, 9, 256)
seq_lengths = [128, 512, 1024, 2048, 4096, 8192, 4096, 2048, 1024, 512, 128]

print('=' * 32)
print('Result by new version')
result_by_rope_ntk_new = [rope_ntk_new(x, seq_length) for seq_length in seq_lengths]
for i in range(len(seq_lengths) // 2):
    cos_before = result_by_rope_ntk_new[i][0]
    cos_after = result_by_rope_ntk_new[len(seq_lengths) - i - 1][0]
    sin_before = result_by_rope_ntk_new[i][1]
    sin_after = result_by_rope_ntk_new[len(seq_lengths) - i - 1][1]
    print(f"length {seq_lengths[i]} {seq_lengths[len(seq_lengths) - i - 1]}")
    print(f"\tcos is equal: {torch.equal(cos_before, cos_after)}")
    print(f"\tsin is equal: {torch.equal(sin_before, sin_after)}")

print('=' * 32)
print('Result by old version')
result_by_rope_ntk_old = [rope_ntk_old(x, seq_length) for seq_length in seq_lengths]
for i in range(len(seq_lengths) // 2):
    cos_before = result_by_rope_ntk_old[i][0]
    cos_after = result_by_rope_ntk_old[len(seq_lengths) - i - 1][0]
    sin_before = result_by_rope_ntk_old[i][1]
    sin_after = result_by_rope_ntk_old[len(seq_lengths) - i - 1][1]
    print(f"length {seq_lengths[i]} {seq_lengths[len(seq_lengths) - i - 1]}")
    print(f"\tcos is equal: {torch.equal(cos_before, cos_after)}")
    print(f"\tsin is equal: {torch.equal(sin_before, sin_after)}")


seq_lengths = [random.randint(1, 32768) for _ in range(10000)]
start = time.time()
for seq_length in seq_lengths:
    rope_ntk_new(x, seq_length)
end = time.time()
print(f"{end - start:.3f}s total, {(end - start) / 10000:.5f}s each")

start = time.time()
for seq_length in seq_lengths:
    rope_ntk_old(x, seq_length)
end = time.time()
print(f"{end - start:.3f}s total, {(end - start) / 10000:.5f}s each")
```
Output:
```shell
================================
Result by new version
length 128 128
	cos is equal: True
	sin is equal: True
length 512 512
	cos is equal: True
	sin is equal: True
length 1024 1024
	cos is equal: True
	sin is equal: True
length 2048 2048
	cos is equal: True
	sin is equal: True
length 4096 4096
	cos is equal: True
	sin is equal: True
================================
Result by old version
length 128 128
	cos is equal: False
	sin is equal: False
length 512 512
	cos is equal: False
	sin is equal: False
length 1024 1024
	cos is equal: False
	sin is equal: False
length 2048 2048
	cos is equal: False
	sin is equal: False
length 4096 4096
	cos is equal: False
	sin is equal: False

7.573s total, 0.00076s each
0.101s total, 0.00001s each
```

### CODE REVIEW

**Comment by i4never** [comment]
Retriggering the test may fix the test timeout.

**Comment by ArthurZucker** [comment]
Yep, rebasing on main should also help!

**Comment by ArthurZucker** [comment]
Ouch 😅

**Comment by i4never** [comment]
Each forward will take several millseconds more. If we cache cos/sin in first layer and pass it to remains (as [llama in tgi](https://github.com/huggingface/text-generation-inference/blob/96a982ad8fc232479384476b1596a880697cc1d0/server/text_generation_server/models/custom_modeling/flash_llama_modeling.py#L443-L460) doing), more time could be saved.

**Comment by ArthurZucker** [comment]
Yeah for sure, but that is a bit too big of a change

File: src/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -283,23 +283,37 @@ def __init__(self, dim, max_position_embeddings=2048, base=10000, device=None, s
...
 
         if seq_len > self.max_position_embeddings:
             base = self.base * (
                 (self.scaling_factor * seq_len / self.max_position_embeddings) - (self.scaling_factor - 1)
             ) ** (self.dim / (self.dim - 2))
-            inv_freq = 1.0 / (base ** (torch.arange(0, self.dim, 2).float().to(device) / self.dim))
-            self.register_buffer("inv_freq", inv_freq, persistent=False)
+        else:
+            base = self.base
+        inv_freq = 1.0 / (base ** (torch.arange(0, self.dim, 2).float().to(device) / self.dim))
```
Comment (gante) [inline src/transformers/models/llama/modeling_llama.py:297]: The fact that this line is not within an if/else could explain part of the slowdown :)

(better to have some duplication than to have such a large slowdown)

File: src/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -283,23 +283,37 @@ def __init__(self, dim, max_position_embeddings=2048, base=10000, device=None, s
...
-        freqs = torch.einsum("i,j->ij", t, self.inv_freq)
+        freqs = torch.einsum("i,j->ij", t, inv_freq)
         # Different from paper, but it uses a different permutation in order to obtain the same calculation
         emb = torch.cat((freqs, freqs), dim=-1)
         self.register_buffer("cos_cached", emb.cos().to(dtype), persistent=False)
         self.register_buffer("sin_cached", emb.sin().to(dtype), persistent=False)
 
+    def forward(self, x, seq_len=None):
+        # x: [bs, num_attention_heads, seq_len, head_size]
+        if not (seq_len <= self.max_seq_len_cached == self.max_position_embeddings):
```
Comment (gante) [inline src/transformers/models/llama/modeling_llama.py:309]: As it is written, this condition is `True` when:
1. seq_len > self.max_seq_len_cached OR
2. self.max_seq_len_cached != self.max_position_embeddings

This means that this is always called except when `seq_len == self.max_seq_len_cached == self.max_position_embeddings`, resulting in a big slowdown.

What we want here is probably 
```
if seq_len > self.max_seq_len_cached or (seq_len < self.max_seq_len_cached and self.max_seq_len_cached > self.max_position_embeddings):
```
This translates to `True` strictly when we need to recompute the sin cos variables, saving precious time!

**Review by gante — COMMENTED** [review]
Hi @i4never, thank you for opening the PR 👋 

As @ArthurZucker mentioned, a 10x slowdown is too much for a precision correction. ~Fortunately, I believe I've identified the major causes for slowdowns :) Check my comments below~

P.S.: something seems to have gone wrong in the rebase, the diff is huge 🤔

File: src/transformers/models/deprecated/open_llama/modeling_open_llama.py [hunk 3]
```diff
@@ -161,23 +161,37 @@ def __init__(self, dim, max_position_embeddings=2048, base=10000, device=None, s
         super().__init__(dim, max_position_embeddings, base, device)
 
     def _set_cos_sin_cache(self, seq_len, device, dtype):
+        if seq_len <= self.max_position_embeddings:
+            seq_len = self.max_position_embeddings
+
```
Comment (i4never) [inline src/transformers/models/deprecated/open_llama/modeling_open_llama.py:166]: @gante `seq_len` is set to `max_position_embeddings` here. Minimal cached length is `max_position_embeddings`. I think `seq_len <= self._seq_len_cached==self.max_position_embeddings` works.

Extra time mainly caused by recompute `max(seq_len - max_position_embeddings, 0) * len(generated_token)` times. Haven't thought of a good way yet.

Comment (gante) [reply src/transformers/models/deprecated/open_llama/modeling_open_llama.py:166]: I see, you're right! I forgot to account for this change :)

**Comment by gante** [comment]
Then only the issue of throughput remains. 

I have no obvious suspect of what could be causing it. I have two suggestions: 
1. as a sanity check, confirm that there is no throughput degradation as you increase the sequence length (and not as you reduce it). The changes you introduced should cause no changes here.
2. Run your whole test script in a loop a few times, and discard the first 2 iterations. The slowdown could be due to GPU initialization time.

**Pull request closed without merging**